File: apps/javascript/design-pattern/src/template-method.js
// Template Method パターン
// アルゴリズムの骨格を基底クラスで定義し、具体的なステップをサブクラスに委ねる

export class Report {
  constructor() {
    this.title = '月次報告';
    this.text = ['順調', '最高の調子'];
  }

  // テンプレートメソッド: レポート出力の骨格
  outputReport() {
    const lines = [];
    lines.push(...this.outputStart());
    lines.push(...this.outputHead());
    lines.push(...this.outputBodyStart());
    lines.push(...this.outputBody());
    lines.push(...this.outputBodyEnd());
    lines.push(...this.outputEnd());
    return lines.join('\n');
  }

  outputBody() {
    return this.text.flatMap((line) => this.outputLine(line));
  }

  // フックメソッド（デフォルトは何もしない）
  outputStart() { return []; }
  outputHead() { return this.outputLine(this.title); }
  outputBodyStart() { return []; }

  // 抽象メソッド
  outputLine(_line) {
    throw new Error('サブクラスで outputLine を実装してください');
  }

  outputBodyEnd() { return []; }
  outputEnd() { return []; }
}

export class HtmlReport extends Report {
  outputStart() { return ['<html>']; }
  outputHead() {
    return [' <head>', ` <title>${this.title}</title>`, ' </head>'];
  }
  outputBodyStart() { return ['<body>']; }
  outputLine(line) { return [` <p>${line}</p>`]; }
  outputBodyEnd() { return ['</body>']; }
  outputEnd() { return ['</html>']; }
}

export class PlainTextReport extends Report {
  outputHead() {
    return [`**** ${this.title} ****`, ''];
  }

  outputLine(line) {
    return [line];
  }
}

File: apps/scala/design-pattern/src/TemplateMethod.scala
// Template Method パターン
// アルゴリズムの骨格を trait で定義し、具体的なステップをサブクラスに委ねる

package designpattern.templatemethod

trait Report:
  val title: String     = "月次報告"
  val text: Seq[String] = Seq("順調", "最高の調子")

  // テンプレートメソッド: レポート出力の骨格
  def outputReport(): String =
    val lines =
      outputStart() ++
        outputHead() ++
        outputBodyStart() ++
        outputBody() ++
        outputBodyEnd() ++
        outputEnd()
    lines.mkString("\n")

  def outputBody(): Seq[String] =
    text.flatMap(line => outputLine(line))

  // フックメソッド（デフォルトは何もしない）
  def outputStart(): Seq[String]     = Seq.empty
  def outputHead(): Seq[String]      = outputLine(title)
  def outputBodyStart(): Seq[String] = Seq.empty

  // 抽象メソッド
  def outputLine(line: String): Seq[String]

  def outputBodyEnd(): Seq[String] = Seq.empty
  def outputEnd(): Seq[String]     = Seq.empty

class HtmlReport extends Report:
  override def outputStart(): Seq[String] = Seq("<html>")

  override def outputHead(): Seq[String] =
    Seq(" <head>", s"  <title>$title</title>", " </head>")

  override def outputBodyStart(): Seq[String]        = Seq("<body>")
  override def outputLine(line: String): Seq[String] = Seq(s"  <p>$line</p>")
  override def outputBodyEnd(): Seq[String]          = Seq("</body>")
  override def outputEnd(): Seq[String]              = Seq("</html>")

class PlainTextReport extends Report:

  override def outputHead(): Seq[String] =
    Seq(s"**** $title ****", "")

  override def outputLine(line: String): Seq[String] = Seq(line)
